### DISCUSSION TITLE
How come Prettier moves only the last array access expression?

### DISCUSSION THREAD

**Discussion by zhao-li** [prettier/prettier#13298] [Q&A]
Hi 👋

I used ESLint's Prettier plugin and it auto-fixed this long line of code:
```
const emittedMessage = wrapper.emitted("update:modelValue")[firstEmittedEventIndex][argumentIndex];
```
to this multiline version:
```
const emittedMessage =
  wrapper.emitted("update:modelValue")[firstEmittedEventIndex][
    argumentIndex
  ];
```

It seems odd that only the last array access expression is pushed to the next line.

Is this the expected behavior or is this a bug?

Thank you for your time 🙏

PS:
I would have thought that something along the following lines would be prescribed:
```
const emittedMessage =
  wrapper.emitted("update:modelValue")
    [firstEmittedEventIndex]
    [argumentIndex]
;
```
or
```
const emittedMessage =
  wrapper
    .emitted("update:modelValue")
    [firstEmittedEventIndex]
    [argumentIndex]
;
```

**Comment by harshagarwalnyu** [comment]
Expected behavior. Prettier's algorithm treats `.emitted("update:modelValue")` as a single "group" (a method call with arguments) and `[firstEmittedEventIndex][argumentIndex]` as a subscript chain on top of that group.

When the full line exceeds the print width, Prettier breaks at the outermost level first. It sees the expression as essentially:

```
wrapper.emitted("update:modelValue")   // method call group
  [firstEmittedEventIndex][argumentIndex]  // subscript chain
```

The subscript accesses are attached to the result of the method call, so Prettier keeps them together and only breaks them onto a new line when needed. The asymmetry you're seeing (one subscript inline, one wrapped) happens because Prettier tries to fit as much as possible before breaking — `[firstEmittedEventIndex]` fits alongside the method call after the break, but adding `[argumentIndex]` would exceed the print width, so only that last one wraps.

Your alternative formatting (each subscript on its own indented line) isn't something Prettier currently produces for this pattern. If this is a common issue in your codebase and you want each subscript on its own line, the practical workaround is to extract intermediate results into named variables:

```ts
const emittedEvents = wrapper.emitted("update:modelValue");
const firstEvent = emittedEvents[firstEmittedEventIndex];
const emittedMessage = firstEvent[argumentIndex];
```

Which also tends to produce clearer test output.

**Comment by Maher-Reven** [comment]
Late, but this is still open and the answer has changed since 2022: **your exact example no longer reproduces.**

On 3.9.6 and on `main`, that line now formats as

```js
const emittedMessage =
  wrapper.emitted("update:modelValue")[firstEmittedEventIndex][argumentIndex];
```

Prettier breaks after the `=` and the whole right-hand side then fits in 80 columns, so neither subscript wraps. So the specific asymmetry you asked about is gone for this input.

The behaviour you were asking about is still there when the right-hand side is too long on its own, which is the case worth explaining:

```js
// input
const c = obj[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa][bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb][cccccccccccc];

// output
const c =
  obj[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa][bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb][
    cccccccccccc
  ];
```

The reason is `printMemberLookup` in [`src/language-js/print/member.js`](https://github.com/prettier/prettier/blob/main/src/language-js/print/member.js). A computed lookup is printed as its own group:

```js
return group([optional, "[", indent([softline, property]), softline, "]"]);
```

Each `[...]` is therefore an *independent* group, and a group breaks only when its own contents do not fit the remaining space on the line. The printer fills left to right, so `[firstEmittedEventIndex]` still fits and stays flat, and the next one does not and breaks. That is the asymmetry — it is not a decision to treat the last subscript specially, it is just the first group that runs out of room.

Two details that follow from the same code, in case they help:

- A non-computed `.foo` lookup returns `[optional, ".", property]` with no group and no softline, so it can never break on its own. That is why `.emitted(...)` stays glued to `wrapper`.
- A numeric subscript such as `[0]` returns `[optional, "[", property, "]"]`, also ungrouped, so it never breaks either.

Both layouts you proposed would need the subscripts to be collected into one shared group so that they break together, which is what [`member-chain.js`](https://github.com/prettier/prettier/blob/main/src/language-js/print/member-chain.js) does for method-call chains but deliberately not for plain computed lookups.
